File: 127-word-ladder/word-ladder.java
class Pair{
    public String word;
    public int step;

    Pair(String word, int step){
        this.word = word;
        this.step = step;
    }
}

class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        HashSet<String> set = new HashSet<>(wordList);
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(beginWord, 1));

        while(!q.isEmpty()){
            Pair top = q.poll();
            String word = top.word;
            int step = top.step;

            if(word.equals(endWord)){
                return step;
            }

            for(int i = 0; i < word.length(); i++){
                char[] charArray = word.toCharArray();
                char original = charArray[i];
                for(char c = 'a'; c <= 'z'; c++){
                    charArray[i] = c;
                    String str = new String(charArray);
                    if(set.contains(str)){
                        set.remove(str);
                        q.offer(new Pair(str, step + 1));
                    }
                }
            }
        }
        return 0;
    }
}

File: 127-word-ladder/word-ladder.py
from collections import deque # Import deque for BFS queue

class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: list) -> int:
        wordSet = set(wordList) # Convert wordList to set for O(1) lookup
        if endWord not in wordSet:
            return 0 # If endWord not in wordList, no transformation possible

        queue = deque() # Initialize BFS queue
        queue.append((beginWord, 1)) # Start with beginWord and step count 1
        visited = set() # Set to track visited words
        visited.add(beginWord) # Mark beginWord as visited

        while queue:
            word, steps = queue.popleft() # Pop word and current step count
            if word == endWord:
                return steps # Found endWord, return steps

            for i in range(len(word)): # For each character position
                for c in 'abcdefghijklmnopqrstuvwxyz': # Try all possible letters
                    nextWord = word[:i] + c + word[i+1:] # Change one letter
                    if nextWord in wordSet and nextWord not in visited:
                        queue.append((nextWord, steps + 1)) # Enqueue new word with incremented steps
                        visited.add(nextWord) # Mark as visited

        return 0 # If endWord not found, return 0